### ISSUE TITLE
August 3 updates in the Next branch cause problems in AMD/Nvidia GPU mixed-use cases?

### ISSUE THREAD

**Issue by iteachcs** [gcp/leela-zero#1701]
A few friends of mine with configurations of one AMD GPU and one Nvidia GPU have experienced issues with the Next branch since the August 3 updates.  Their AMD and Nvidia GPUs were able to work together using the compiled Next branch up to July 30.  

I was wondering if it is possible that the changes on August 3 were not compatible with these mixed-use cases.  It is interesting that the GPUs do work individually using the compiled August 3 version.  The problem just happens when they try to use the two different GPUs together.

**Comment by roy7** [comment]
Maybe they can try "--precision single" and see if that fixes it?

**Comment by iteachcs** [comment]
Thanks for the answer!  The mixed use GPUs can work with the "--precision single" option.  

I am asking my friend to test a little bit further to see if Leelaz works with the "--precision half" option with only the AMD GPU.  I will report back what I find.

**Comment by iteachcs** [comment]
We did some further testing.  It appears that the problem is actually caused by the AMD GPU.  

If we use the AMD GPU alone with the "--precision half" or "use-half" (July 30 version) option, Leelaz will crash.  However, there is no problem if we use the Nvidia GPU alone with the same parameters.  

We believe that the reason why the single AMD GPU worked when we tested last time was that we did not use the "--precision half" option.  The speedup was not big enough for "half precision" to kick in.  Once we explicitly required "half precision"  with the "--precision half" option, the problem appeared.

In summary, the problem is not about mixed use.  It is that the AMD GPU supports neither FP16 Storage nor FP16 Computer.

I will leave this issue open for a while to see if there is any further input.

**Comment by y-ich** [comment]
Hi.

I tried to use next branch before August 3rd and after August 3rd on RX Vega 56 with ROCm driver, which supports half float acceleration.

My experiences were,

Before August 3rd:
Using 62b weight(ELF OpenGo v0), it did not work. The tuner at first finished anyway but the NN output seemed kind of nosense.
Using amdgpu-pro driver, which does not support half float acceleration, instead of ROCm driver, it worked.

After August 3rd:
Using 62b weight(ELF OpenGo v0), it works except for the very first calculation of NN for first move. (The value of the first calculation of value network was 1.0)
Using best-network, It failed to tune SGEMM.

    Started OpenCL SGEMM tuner.
    Will try 290 valid configurations.
    Failed to find a working configuration.
    Check your OpenCL drivers.
    terminate called after throwing an instance of 'std::runtime_error'
      what():  Tuner failed to find working configuration.
    Aborted (core dumped)

It depends on what weight you use.

Hope that this information is helpful to resolve this issue.

**Comment by gcp** [comment]
I guess we need to detect this during tuning and fall back to fp32 support, as pointed out in the pull with tuner fixes.

**Comment by gcp** [comment]
@iteachcs Can you test this against the next branch now?

**Comment by iteachcs** [comment]
@gcp My friend tested and said that the newest next branch worked with his AMD GPU card without the need to specify "--precision single".  The program did not work if "--precision half" is set.

**Comment by Marcin1960** [comment]
@iteachcs 

I had to specify "--precision single" for my AMD. Perhaps because I use 32 bit platform.

**Comment by alreadydone** [comment]
@Marcin1960 Did you test with latest gcp/leela-zero/next? Or you can test with my latest komi+next branch which has gcp/leela-zero/next merged in.
